---
title: "Update analysis on the two samples d4_LPS d4_SPF: Ambient RNA clusters only"
author:
  - "trnguyen@ukaachen.de"
date: "Last update on `r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    number_sections: true
    df_print: paged
    toc: true
    toc_float:
      toc_collapsed: false
    toc_depth: 3
    theme: lumen
params:
  outdir: NA
  PROJECT: NA
  s.obj.name: NA
  sub.cluster.id: NA
---
  
```{css zoom-lib-src, echo = FALSE}
script src = "https://ajax.googleapis.com/ajax/libs/jquery/3.4.1/jquery.min.js"
```

```{js zoom-jquery, echo = FALSE}
$(document).ready(function() {
  $('body').prepend('<div class=\"zoomDiv\"><img src=\"\" class=\"zoomImg\"></div>');
  // onClick function for all plots (img's)
    $('img:not(.zoomImg)').click(function() {
      $('.zoomImg').attr('src', $(this).attr('src')).css({width: '100%'});
      $('.zoomDiv').css({opacity: '1', width: 'auto', border: '1px solid white', borderRadius: '5px', position: 'fixed', top: '50%', left: '50%', marginRight: '-50%', transform: 'translate(-50%, -50%)', boxShadow: '0px 0px 50px #888888', zIndex: '50', overflow: 'auto', maxHeight: '100%'});
    });
    // onClick function for zoomImg
    $('img.zoomImg').click(function() {
      $('.zoomDiv').css({opacity: '0', width: '0%'}); 
    });
  });
```

<style type="text/css">
    div.datatables { height: auto !important;}
</style>

```{r echo=FALSE, warning=FALSE, message=FALSE, results='hide', fig.width=20, fig.height=12}
#####----------------------------------------------------------------------#####
#
# trnguyen@ukaachen.de
#
#####----------------------------------------------------------------------#####
path.to.main.src <- "/home/hieunguyen/CRC1382/src_2023/EStange/official_SeuratV4"
scrna_pipeline_src <- "/home/hieunguyen/CRC1382/src_2023/src_pipeline/scRNA_GEX_pipeline/processes_src"
source(file.path(scrna_pipeline_src, "import_libraries.R"))
source(file.path(scrna_pipeline_src, "helper_functions.R"))

#####----------------------------------------------------------------------#####
# CONFIGURATIONS 
#####----------------------------------------------------------------------#####
analysis.round <- "1st"

chosen.seed <- 42
num.dim.integration <- 25 
num.PCA <- 25
num.dim.cluster <- 25
num.PC.used.in.Clustering <- 25

sample1 <- "d4_LPS"
sample2 <- "d4_SPF"

if ("heatmaply" %in% installed.packages() == FALSE){
  install.packages("heatmaply")
}
path.to.src <- "/home/hieunguyen/CRC1382/src/src_pipeline/scRNA_GEX_pipeline/processes_src"

outdir <- "/home/hieunguyen/CRC1382/outdir"
PROJECT <- "EStange_20240411_SeuratV4_reduced_RNAcontam_0"
sub.cluster.id <- "Myeloid_Basophils"
s.obj.name <- "all_sobj.integrated.rds"

# outdir <- params$outdir
# PROJECT <- params$PROJECT
# sub.cluster.id <- params$sub.cluster.id
# s.obj.name <- params$s.obj.name

path.to.main.output <- file.path(outdir, PROJECT, "data_analysis")

path.to.01.output <- file.path(path.to.main.output, "01_output")
path.to.02.output <- file.path(path.to.main.output, "02_output")
path.to.03.output <- file.path(path.to.main.output, "03_output")
path.to.08.output <- file.path(path.to.main.output, "08_output", sub.cluster.id)
path.to.09.output <- file.path(path.to.main.output, "09_output", sub.cluster.id)
dir.create(path.to.09.output, showWarnings = FALSE, recursive = TRUE)
s.obj <- readRDS(file.path(path.to.08.output, sprintf("%s.remove_contam_T_B_cells.ambientRNAonly.rds", sub.cluster.id)))
```

## UMAP after removing T/B contaminated cells
```{r echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.width=14, fig.height=10}
DimPlot(object = s.obj, reduction = "INTE_UMAP", label = TRUE, label.box = TRUE, repel = TRUE, group.by = "celltype")
```

## Ambient RNA levels on UMAP
```{r echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.width=20, fig.height=10}
FeaturePlot(object = s.obj, reduction = "INTE_UMAP", features = "AmbientRNA", split.by = "name")
```

## Ambient RNA levels, violin plot
```{r echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.width=20, fig.height=10}
VlnPlot(object = s.obj, features = "AmbientRNA", split.by = "name")
```

# Compare UMI and RNA counts between d4_LPS and d4_SPF

Note Total RNA count per clusters and additionally in ambient RNA contaminated Clusters* in comparison between d4_LPS vs. d4_SPF

- `nFeature_RNA`: Number of genes that have > 0 reads in each cell. 

- `nCount_RNA`: number of reads in each cell. 

## nCount_RNA {.tabset}
### UMAP
```{r echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.width=20, fig.height=10}
FeaturePlot(object = s.obj, split.by = "name", features = c("nCount_RNA"), label = TRUE, label.size = 8, reduction = "INTE_UMAP")
```

### Violin plot
```{r echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.width=20, fig.height=10}
VlnPlot(object = s.obj, features = c("nCount_RNA"), split.by = "name")
```

## nFeature_RNA {.tabset}
### UMAP
```{r echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.width=20, fig.height=10}
FeaturePlot(object = s.obj, split.by = "name", features = c("nFeature_RNA"), label = TRUE, label.size = 8, reduction = "INTE_UMAP")
```

### Violin plot
```{r echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.width=20, fig.height=10}
VlnPlot(object = s.obj, features = c("nFeature_RNA"), split.by = "name")
```

# Show cells from "contamninated ambient RNA" clusters

```{r echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.width=20, fig.height=10}
ambientRNA.cells <- colnames(subset(s.obj, ambientRNA.cluster == "yes"))
DimPlot(object = s.obj, reduction = "INTE_UMAP", cells.highlight = ambientRNA.cells, split.by = "name")
```

# Cluster marker genes
```{r echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.width=25, fig.height=12}
if (file.exists(file.path(path.to.09.output, sprintf("cluster_markers_subclusters_%s.rds", sub.cluster.id))) == FALSE){
  cluster.markers <- FindAllMarkers(object = s.obj, test.use = "wilcox", assay = "RNA")
  cluster.markers <- subset(cluster.markers, cluster.markers$p_val_adj <= 0.05 & cluster.markers$avg_log2FC > 0)
  saveRDS(cluster.markers, file.path(path.to.09.output, sprintf("cluster_markers_subclusters_%s.rds", sub.cluster.id)))
} else {
  cluster.markers <- readRDS(file.path(path.to.09.output, sprintf("cluster_markers_subclusters_%s.rds", sub.cluster.id)))
}
```

## Tables: All cluster marker genes {.tabset}

```{r echo=FALSE, warning=FALSE, results='asis', include=FALSE}
cluster.markers %>% create_dt()
```

```{r echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.width=20, fig.height=12}
for(cluster.id in sort(unique(cluster.markers$cluster))){
  cat(paste("\n\n### Cluster: ", cluster.id, "##\n"))
  tmp <- subset(cluster.markers, cluster.markers$cluster == cluster.id) %>% arrange(desc(avg_log2FC)) %>%
    mutate_if(is.numeric, round, 6)
  row.names(tmp) <- NULL
  print( htmltools::tagList(datatable(tmp, extensions = 'Buttons',
                filter = "top",
                options = list(dom = 'Blfrtip',
                               buttons = c('copy', 'csv', 'excel', 'pdf', 'print'),
                               lengthMenu = list(c(10,25,50,-1),
                                                 c(10,25,50,"All")),
                               columnDefs = list(list(
                                 targets = "_all",
                                 render = JS(
                                   "function(data, type, row, meta) {",
                                   "return type === 'display' && data != null && data.length > 100 ?",
                                   "'<span title=\"' + data + '\">' + data.substr(0, 100) + '...</span>' : data;",
                                   "}")
))
))))
cat("\n \n")
}
```

## Top-16 cluster markers feature plot {.tabset}

```{r results='asis', echo=FALSE, message=FALSE, warning=FALSE, fig.width=12, fig.height=8}
for (cluster.id in unique(cluster.markers$cluster)){
  DefaultAssay(s.obj) <- "RNA"
  tmp <- subset(cluster.markers, cluster.markers$cluster == cluster.id & cluster.markers$p_val_adj < 0.05) %>% 
    rowwise %>% 
    mutate_if(is.numeric, round, 6) %>% 
    arrange(desc(avg_log2FC)) %>% head(16)
  cat(sprintf("### Cluster %s \n", cluster.id))
  p <- FeaturePlot(object = s.obj, reduction = "INTE_UMAP", features = tmp$gene, ncol = 4, ) 
  print(p)
  cat("\n \n")
}
```

## Top-16 cluster markers dot plot {.tabset}

```{r results='asis', echo=FALSE, message=FALSE, warning=FALSE, fig.width=12, fig.height=8}
for (cluster.id in unique(cluster.markers$cluster)){
  DefaultAssay(s.obj) <- "RNA"
  tmp <- subset(cluster.markers, cluster.markers$cluster == cluster.id & cluster.markers$p_val_adj < 0.05) %>% 
    rowwise %>% 
    mutate_if(is.numeric, round, 6) %>% 
    arrange(desc(avg_log2FC)) %>% head(16)
  cat(sprintf("### Cluster %s \n", cluster.id))
  p <- DotPlot(object = s.obj, features = tmp$gene) 
  print(p)
  cat("\n \n")
}
```

## Top-16 cluster markers Violin plot {.tabset}

```{r results='asis', echo=FALSE, message=FALSE, warning=FALSE, fig.width=12, fig.height=8}
for (cluster.id in unique(cluster.markers$cluster)){
  DefaultAssay(s.obj) <- "RNA"
  tmp <- subset(cluster.markers, cluster.markers$cluster == cluster.id & cluster.markers$p_val_adj < 0.05) %>% 
    rowwise %>% 
    mutate_if(is.numeric, round, 6) %>% 
    arrange(desc(avg_log2FC)) %>% head(16)
  cat(sprintf("### Cluster %s \n", cluster.id))
  p <- VlnPlot(object = s.obj, features = tmp$gene, pt.size = 0) 
  print(p)
  cat("\n \n")
}
```


## Heatmap top-5 up regulated genes in each cluster
```{r results='asis', echo=FALSE, message=FALSE, warning=FALSE, fig.width=28, fig.height=20}
topN <- 5
tmp.cluster.markers.up <- subset(cluster.markers, cluster.markers$avg_log2FC >= 0 & cluster.markers$p_val_adj <= 0.05)

selected.genes <- data.frame()
for (cluster.id in unique(tmp.cluster.markers.up$cluster)){
  tmp <- subset(tmp.cluster.markers.up, tmp.cluster.markers.up$cluster == cluster.id) %>% arrange(desc(avg_log2FC))
  tmp.genes <- head(tmp, topN)
  selected.genes <- rbind(selected.genes, tmp.genes)
}

p <- DoHeatmap(object = s.obj, features = selected.genes$gene) + scale_fill_viridis_c(option = "plasma")
print(p)
```

## Number of cells per cluster
```{r results='asis', echo=FALSE, message=FALSE, warning=FALSE, fig.width=28, fig.height=20}
s.obj@meta.data %>% subset(select = c(seurat_clusters)) %>% table() %>% as.data.frame() %>% create_dt()
```


# Ambient RNA contamination on UMAP 

## All samples, integrated
```{r echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.width=14, fig.height=10}
FeaturePlot(object = s.obj, features = "AmbientRNA", label = TRUE, label.size = 8, reduction = "INTE_UMAP")
```

## Each sample, integrated UMAP {.tabset}
```{r echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.width=14, fig.height=10}
for (sample.id in unique(s.obj$name)){
  cat(sprintf("### Sample %s \n", sample.id))
  p <- FeaturePlot(object = subset(s.obj, name == sample.id), features = "AmbientRNA", label = TRUE, label.size = 8, reduction = "INTE_UMAP")
  print(p)
  cat("\n \n")
}
```

## Each sample, Violin plot ambient RNA contamination level in each cluster {.tabset}
```{r echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.width=14, fig.height=10}
for (sample.id in unique(s.obj$name)){
  cat(sprintf("### Sample %s \n", sample.id))
  p <- VlnPlot(object = subset(s.obj, name == sample.id), features = "AmbientRNA")
  print(p)
  cat("\n \n")
}
```

# Differential gene expression analysis: high ambient RNA contaminated cells vs others

- positive logFC indicates that the gene is higher in the "ambient RNA contaminated cells >= 50%" cells. - negative logFC indicates that the gene is lower in the "ambient RNA contaminated cells >= 50%" cells.

```{r echo=FALSE, fig.height=10, fig.width=14, message=FALSE, warning=FALSE, results='asis'}
high.ambientRNA.cells <- subset(s.obj@meta.data, s.obj@meta.data$AmbientRNA >= 0.5) %>% row.names()

meta.data <- s.obj@meta.data %>% rownames_to_column("barcode") %>%
  rowwise() %>%
  mutate(high.ambientRNA = ifelse(barcode %in% high.ambientRNA.cells, "yes", "no")) %>%
  column_to_rownames("barcode")
meta.data <- meta.data[row.names(s.obj@meta.data), ]
s.obj <- AddMetaData(object = s.obj, metadata = meta.data$high.ambientRNA, col.name = "high.ambientRNA")

if (file.exists(file.path(path.to.09.output, sprintf("high.ambientRNA.markers_%s.rds", sub.cluster.id))) == FALSE){
  high.ambientRNA.markers <- FindMarkers(object = s.obj, ident.1 = "yes", ident.2 = "no", group.by = "high.ambientRNA", assay = "RNA", test.use = "wilcox")
  high.ambientRNA.markers.raw <- high.ambientRNA.markers
  high.ambientRNA.markers <- subset(high.ambientRNA.markers, high.ambientRNA.markers$p_val_adj <= 0.05)
  saveRDS(high.ambientRNA.markers.raw, file.path(path.to.09.output, sprintf("high.ambientRNA.markers_%s.raw.rds", sub.cluster.id)))
  saveRDS(high.ambientRNA.markers, file.path(path.to.09.output, sprintf("high.ambientRNA.markers_%s.rds", sub.cluster.id)))
} else {
  high.ambientRNA.markers <- readRDS(file.path(path.to.09.output, sprintf("high.ambientRNA.markers_%s.rds", sub.cluster.id)))
  high.ambientRNA.markers.raw <- readRDS(file.path(path.to.09.output, sprintf("high.ambientRNA.markers_%s.raw.rds", sub.cluster.id)))
}

high.ambientRNA.markers <- high.ambientRNA.markers %>% rownames_to_column("Gene") %>% rowwise() %>%
  mutate(abs.avg_log2FC = abs(avg_log2FC)) %>% arrange(desc(abs.avg_log2FC))
```

## Full table: All differentially expressed genes
```{r results='asis', echo=FALSE, message=FALSE, warning=FALSE, fig.width=14, fig.height=10}
high.ambientRNA.markers %>% create_dt()
```

```{r results='asis', echo=FALSE, message=FALSE, warning=FALSE, fig.width=14, fig.height=10}
input.df <- high.ambientRNA.markers.raw %>% rownames_to_column("Gene")
all.pvals <- unique(input.df$p_val_adj)
all.pvals <- all.pvals[all.pvals != 0]
min.pval.adj <- min(all.pvals)
input.df <- input.df %>% rowwise() %>%
  mutate(p_val_adj = ifelse(p_val_adj == 0, min.pval.adj, p_val_adj))

top10up.genes <- input.df %>% subset(p_val_adj <= 0.05 & avg_log2FC > 0) %>% arrange(desc(avg_log2FC)) %>% head(10)
top10down.genes <- input.df %>% subset(p_val_adj <= 0.05 & avg_log2FC < 0) %>% arrange(desc(avg_log2FC)) %>% tail(10)
input.df <- input.df %>%
rowwise() %>%
mutate(sig = ifelse(p_val_adj <= 0.05, "Sig.", "not Sig.")) %>%
  mutate(show.gene.name = ifelse(Gene %in% c(top10up.genes$Gene, top10down.genes$Gene), Gene, NA))
  
cutoff.adjp <- 0.05
volcano.plot <- ggplot(data=input.df, aes(x=avg_log2FC, y=-log10(p_val_adj ), col = sig, label=Gene)) +
  geom_point() +
  geom_label_repel(label = input.df$show.gene.name, size = 8, max.overlaps = 100) +
  scale_color_manual(values=c("#c0d2f0", "#f28095")) +
  theme_minimal() +
  geom_vline(xintercept=c(-1, 1), col="#9a9fa6", linetype='dotted') +
  geom_hline(yintercept=-log10(cutoff.adjp), col="#9a9fa6", linetype='dotted') +
  theme_bw() +
  theme(plot.title = element_text(hjust=0.5, face="bold", size = 12), axis.text=element_text(size=12))
volcano.plot
```
