---
title: "Sub-clustering `r params$sub.cluster.id`"
author:
  - "trnguyen@ukaachen.de"
date: "Last update on `r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    keep_md: no
    df_print: paged
    toc: true
    toc_float:
      toc_collapsed: false
    toc_depth: 3
    number_sections: true
    theme: lumen
params:
  outdir: NA
  PROJECT: NA
  sub.cluster.id: NA
  s.obj.name: NA
---

  
  
```{css zoom-lib-src, echo = FALSE}
script src = "https://ajax.googleapis.com/ajax/libs/jquery/3.4.1/jquery.min.js"
```

```{js zoom-jquery, echo = FALSE}
$(document).ready(function() {
  $('body').prepend('<div class=\"zoomDiv\"><img src=\"\" class=\"zoomImg\"></div>');
  // onClick function for all plots (img's)
    $('img:not(.zoomImg)').click(function() {
      $('.zoomImg').attr('src', $(this).attr('src')).css({width: '100%'});
      $('.zoomDiv').css({opacity: '1', width: 'auto', border: '1px solid white', borderRadius: '5px', position: 'fixed', top: '50%', left: '50%', marginRight: '-50%', transform: 'translate(-50%, -50%)', boxShadow: '0px 0px 50px #888888', zIndex: '50', overflow: 'auto', maxHeight: '100%'});
    });
    // onClick function for zoomImg
    $('img.zoomImg').click(function() {
      $('.zoomDiv').css({opacity: '0', width: '0%'}); 
    });
  });
```

<style type="text/css">
    div.datatables { height: auto !important;}
</style>


```{r echo=FALSE, warning=FALSE, message=FALSE, results='hide', fig.width=20, fig.height=12}
#####----------------------------------------------------------------------#####
#
# trnguyen@ukaachen.de
#
#####----------------------------------------------------------------------#####
scrna_pipeline_src <- "/home/hieunguyen/CRC1382/src_2023/src_pipeline/scRNA_GEX_pipeline/processes_src"
source(file.path(scrna_pipeline_src, "import_libraries.R"))
source(file.path(scrna_pipeline_src, "helper_functions.R"))

#####----------------------------------------------------------------------#####
# CONFIGURATIONS 
#####----------------------------------------------------------------------#####
analysis.round <- "1st"

chosen.seed <- 42
num.dim.integration <- 25 
num.PCA <- 25
num.dim.cluster <- 25
num.PC.used.in.Clustering <- 25

path.to.src <- "/home/hieunguyen/CRC1382/src/src_pipeline/scRNA_GEX_pipeline/processes_src"

# outdir <- "/home/hieunguyen/CRC1382/outdir"
# PROJECT <- "EStange_20240411_SeuratV4_reduced_RNAcontam_0"
# s.obj.name <- "all_sobj.integrated.rds"
# sub.cluster.id <- "Myeloid_Basophils"

outdir <- params$outdir
PROJECT <- params$PROJECT
s.obj.name <- params$s.obj.name
sub.cluster.id <- params$sub.cluster.id

path.to.main.output <- file.path(outdir, PROJECT, "data_analysis")

path.to.01.output <- file.path(path.to.main.output, "01_output")
path.to.02.output <- file.path(path.to.main.output, "02_output")
path.to.03.output <- file.path(path.to.main.output, "03_output")
path.to.04.output <- file.path(path.to.main.output, "04_output", s.obj.name)
dir.create(path.to.04.output, showWarnings = FALSE, recursive = TRUE)

s.obj.original <- readRDS(file.path(path.to.03.output, s.obj.name))
```

# UMAP
## Original dataset
```{r echo=FALSE, warning=FALSE, message=FALSE, results='hide', fig.width=20, fig.height=12}
DimPlot(object = s.obj.original, reduction = "INTE_UMAP", label = TRUE, label.box = TRUE, repel = TRUE)
```

## Selected clusters
```{r echo=FALSE, warning=FALSE, message=FALSE, results='hide', fig.width=20, fig.height=12}
sub.clusters <- hash()
sub.clusters[["all_sobj.integrated.rds"]] <- list(
  Myeloid_Basophils = c(20, 5, 3, 8, 17, 16, 25, 26, 9, 7, 12, 28, 11),
  B_cells = c(2, 21, 27, 29, 15, 10, 22),
  T_cells = c(14, 1, 6, 13, 4, 24, 0))

DimPlot(object = subset(s.obj.original, seurat_clusters %in% sub.clusters[[s.obj.name]][[sub.cluster.id]]), 
        reduction = "INTE_UMAP", label = TRUE, label.box = TRUE, repel = TRUE)
```

## Group by annotated subclusters

```{r echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.width=20, fig.height=12}
dir.create(file.path(path.to.04.output, "preprocessed_subcluster_obj"), showWarnings = FALSE, recursive = TRUE)

if (file.exists(file.path(path.to.04.output, "preprocessed_subcluster_obj", sprintf("preprocessed_subclusters_%s.rds", sub.cluster.id))) == FALSE){
  my_random_seed <- 42
  
  num.dim.integration <- 25 
  num.PCA <- 25
  num.dim.cluster <- 25
  num.PC.used.in.Clustering <- 25
  num.PC.used.in.UMAP <- 25
  cluster.resolution  <- 0.5
  
  s.obj <- subset(s.obj.original, seurat_clusters %in% sub.clusters[["all_sobj.integrated.rds"]][[sub.cluster.id]])
  s.obj <- DietSeurat(s.obj)
  
  chosen.assay <- "RNA"
  DefaultAssay(s.obj) <- chosen.assay
  
  ##### Re run UMAP and integration
  s.obj <- NormalizeData(s.obj) # ---> use Log Normalized
  s.obj <- FindVariableFeatures(s.obj, selection.method = "vst")
  s.obj <- ScaleData(s.obj, features = rownames(s.obj))
    
  s.obj <- RunPCA(s.obj, npcs = num.PCA, verbose = FALSE, reduction.name=sprintf("%s_PCA", chosen.assay))
    
  s.obj <- RunUMAP(s.obj, reduction = sprintf("%s_PCA", chosen.assay), 
                     dims = 1:num.PC.used.in.UMAP, reduction.name=sprintf("%s_UMAP", chosen.assay),
                     seed.use = my_random_seed, umap.method = "uwot")
    
  # clustering 
  s.obj <- FindNeighbors(s.obj, reduction = sprintf("%s_PCA", chosen.assay), dims = 1:num.PC.used.in.Clustering)
    
  s.obj <- FindClusters(s.obj, resolution = cluster.resolution, random.seed = 0)
  
  #### Integration
  data.list <- SplitObject(s.obj, split.by = "name")
  data.list <- lapply(X = data.list, FUN = function(x) {
      x <- NormalizeData(x)
      x <- FindVariableFeatures(x, selection.method = "vst", nfeatures = 2000)})
    
  k.filter <- 200
    
  anchors <- FindIntegrationAnchors(object.list = data.list, dims = 1:num.dim.integration, scale=F,
                                      k.filter = k.filter) ## THIS IS CCA DIMENSIONS
    
  s.obj_inte <- IntegrateData(anchorset = anchors, dims = 1:num.dim.integration, k.weight = k.filter) ## THIS IS PCA DIMENSION
    
  ## keep the order of integration obj
  s.obj_inte <- s.obj_inte[, colnames(s.obj)]
    
  s.obj[['integrated']] <- s.obj_inte[['integrated']]
    
  s.obj@commands <- c(s.obj@commands, s.obj_inte@commands)
    
  s.obj@tools <- c(s.obj@tools, s.obj_inte@tools)
    
  DefaultAssay(s.obj) <- "integrated"
  s.obj <- ScaleData(s.obj, verbose = FALSE, features = row.names(s.obj))
    
  s.obj <- RunPCA(s.obj, npcs = num.PCA, verbose = FALSE, reduction.name="INTE_PCA")
    
  s.obj <- RunUMAP(s.obj, reduction = "INTE_PCA", dims = 1:num.PCA, reduction.name="INTE_UMAP")
    
    # clustering 
  s.obj <- FindNeighbors(s.obj, reduction = "INTE_PCA", dims = 1:num.dim.cluster)
    
  s.obj <- FindClusters(s.obj, resolution = cluster.resolution)
  saveRDS(s.obj, file.path(path.to.04.output, "preprocessed_subcluster_obj", sprintf("preprocessed_subclusters_%s.rds", sub.cluster.id)))
  
} else {
  s.obj <- readRDS(file.path(path.to.04.output, "preprocessed_subcluster_obj", sprintf("preprocessed_subclusters_%s.rds", sub.cluster.id)))
}

```
 

```{r echo=FALSE, warning=FALSE, message=FALSE, results='hide', fig.width=20, fig.height=12}
##### CONFIGURATIONS
meta.data <- s.obj@meta.data %>% rownames_to_column("barcode") %>% subset(select = c(barcode, seurat_clusters))

embeddings <- s.obj@reductions$INTE_UMAP@cell.embeddings %>% as.data.frame() %>% rownames_to_column("barcode")
embeddings <- merge(embeddings, meta.data, by.x = "barcode", by.y = "barcode")

get_x_y_for_cluster <- function(cluster.id){
  tmp.embeddings <- subset(embeddings, embeddings$seurat_clusters == cluster.id)
  x <- mean(tmp.embeddings$inte_umap_1)
  y <- mean(tmp.embeddings$inte_umap_2)
  return(c(x, y))
}

x <- to_vec(for (item in seq(0, length(unique(s.obj@meta.data$seurat_clusters))- 1)) get_x_y_for_cluster(item)[[1]])
y <- to_vec(for (item in seq(0, length(unique(s.obj@meta.data$seurat_clusters))- 1)) get_x_y_for_cluster(item)[[2]])

sample_colors <- c("#cb42f5", "#42f584")
num.clusters <- length(unique(s.obj$seurat_clusters))
colors <- c(hue_pal()(num.clusters), sample_colors)
names(colors) <- c(to_vec(for (item in seq(0, num.clusters - 1)) sprintf("%s", item)), c("adult", "d7"))

```

## Selected clusters
```{r echo=FALSE, warning=FALSE, message=FALSE, results='hide', fig.width=20, fig.height=12}
DimPlot(object = subset(s.obj.original, cells = colnames(s.obj)), label = TRUE, label.box = TRUE, repel = TRUE)
```

## UMAP: after re-clustering, group by clusters
```{r echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.width=20, fig.height=12}
p <- DimPlot(object = s.obj, reduction = "INTE_UMAP", label = TRUE, label.box = TRUE, repel = TRUE)
print(p)
```

## UMAP, after re-clustering, group by samples
```{r echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.width=20, fig.height=12}
DimPlot(object = s.obj, reduction = "INTE_UMAP", label = TRUE, label.box = TRUE, repel = TRUE, group.by = "name")
```

# Cluster marker genes
```{r echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.width=25, fig.height=12}
if (file.exists(file.path(path.to.04.output, "preprocessed_subcluster_obj", sprintf("cluster_markers_subclusters_%s.rds", sub.cluster.id))) == FALSE){
  cluster.markers <- FindAllMarkers(object = s.obj, test.use = "wilcox", assay = "RNA")
  cluster.markers <- subset(cluster.markers, cluster.markers$p_val_adj <= 0.05 & cluster.markers$avg_log2FC > 0)
  saveRDS(cluster.markers, file.path(path.to.04.output, "preprocessed_subcluster_obj", sprintf("cluster_markers_subclusters_%s.rds", sub.cluster.id)))
} else {
  cluster.markers <- readRDS(file.path(path.to.04.output, "preprocessed_subcluster_obj", sprintf("cluster_markers_subclusters_%s.rds", sub.cluster.id)))
}
```

## Tables: All cluster marker genes {.tabset}

```{r echo=FALSE, warning=FALSE, results='asis', include=FALSE}
cluster.markers %>% create_dt()
```

```{r echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.width=20, fig.height=12}
for(cluster.id in sort(unique(cluster.markers$cluster))){
  cat(paste("\n\n### Cluster: ", cluster.id, "##\n"))
  tmp <- subset(cluster.markers, cluster.markers$cluster == cluster.id) %>% arrange(desc(avg_log2FC)) %>%
    mutate_if(is.numeric, round, 6)
  row.names(tmp) <- NULL
  print( htmltools::tagList(datatable(tmp, extensions = 'Buttons',
                filter = "top",
                options = list(dom = 'Blfrtip',
                               buttons = c('copy', 'csv', 'excel', 'pdf', 'print'),
                               lengthMenu = list(c(10,25,50,-1),
                                                 c(10,25,50,"All")),
                               columnDefs = list(list(
                                 targets = "_all",
                                 render = JS(
                                   "function(data, type, row, meta) {",
                                   "return type === 'display' && data != null && data.length > 100 ?",
                                   "'<span title=\"' + data + '\">' + data.substr(0, 100) + '...</span>' : data;",
                                   "}")
))
))))
cat("\n \n")
}
```

## Top-16 cluster markers feature plot {.tabset}

```{r results='asis', echo=FALSE, message=FALSE, warning=FALSE, fig.width=12, fig.height=8}
for (cluster.id in unique(cluster.markers$cluster)){
  DefaultAssay(s.obj) <- "RNA"
  tmp <- subset(cluster.markers, cluster.markers$cluster == cluster.id & cluster.markers$p_val_adj < 0.05) %>% 
    rowwise %>% 
    mutate_if(is.numeric, round, 6) %>% 
    arrange(desc(avg_log2FC)) %>% head(16)
  cat(sprintf("### Cluster %s \n", cluster.id))
  p <- FeaturePlot(object = s.obj, reduction = "INTE_UMAP", features = tmp$gene, ncol = 4, ) 
  print(p)
  cat("\n \n")
}
```

## Top-16 cluster markers dot plot {.tabset}

```{r results='asis', echo=FALSE, message=FALSE, warning=FALSE, fig.width=12, fig.height=8}
for (cluster.id in unique(cluster.markers$cluster)){
  DefaultAssay(s.obj) <- "RNA"
  tmp <- subset(cluster.markers, cluster.markers$cluster == cluster.id & cluster.markers$p_val_adj < 0.05) %>% 
    rowwise %>% 
    mutate_if(is.numeric, round, 6) %>% 
    arrange(desc(avg_log2FC)) %>% head(16)
  cat(sprintf("### Cluster %s \n", cluster.id))
  p <- DotPlot(object = s.obj, features = tmp$gene) 
  print(p)
  cat("\n \n")
}
```

## Top-16 cluster markers Violin plot {.tabset}

```{r results='asis', echo=FALSE, message=FALSE, warning=FALSE, fig.width=12, fig.height=8}
for (cluster.id in unique(cluster.markers$cluster)){
  DefaultAssay(s.obj) <- "RNA"
  tmp <- subset(cluster.markers, cluster.markers$cluster == cluster.id & cluster.markers$p_val_adj < 0.05) %>% 
    rowwise %>% 
    mutate_if(is.numeric, round, 6) %>% 
    arrange(desc(avg_log2FC)) %>% head(16)
  cat(sprintf("### Cluster %s \n", cluster.id))
  p <- VlnPlot(object = s.obj, features = tmp$gene, pt.size = 0) 
  print(p)
  cat("\n \n")
}
```


## Heatmap top-5 up regulated genes in each cluster
```{r results='asis', echo=FALSE, message=FALSE, warning=FALSE, fig.width=28, fig.height=20}
topN <- 5
tmp.cluster.markers.up <- subset(cluster.markers, cluster.markers$avg_log2FC >= 0 & cluster.markers$p_val_adj <= 0.05)

selected.genes <- data.frame()
for (cluster.id in unique(tmp.cluster.markers.up$cluster)){
  tmp <- subset(tmp.cluster.markers.up, tmp.cluster.markers.up$cluster == cluster.id) %>% arrange(desc(avg_log2FC))
  tmp.genes <- head(tmp, topN)
  selected.genes <- rbind(selected.genes, tmp.genes)
}

p <- DoHeatmap(object = s.obj, features = selected.genes$gene) + scale_fill_viridis_c(option = "plasma")
print(p)
```

## Number of cells per cluster
```{r results='asis', echo=FALSE, message=FALSE, warning=FALSE, fig.width=28, fig.height=20}
s.obj@meta.data %>% subset(select = c(seurat_clusters)) %>% table() %>% as.data.frame() %>% create_dt()
```


# Ambient RNA contamination on UMAP 

## All samples, integrated
```{r echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.width=14, fig.height=10}
FeaturePlot(object = s.obj, features = "AmbientRNA", label = TRUE, label.size = 8, reduction = "INTE_UMAP")
```

## Each sample, integrated UMAP {.tabset}
```{r echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.width=14, fig.height=10}
for (sample.id in unique(s.obj$name)){
  cat(sprintf("### Sample %s \n", sample.id))
  p <- FeaturePlot(object = subset(s.obj, name == sample.id), features = "AmbientRNA", label = TRUE, label.size = 8, reduction = "INTE_UMAP")
  print(p)
  cat("\n \n")
}
```

## Each sample, Violin plot ambient RNA contamination level in each cluster {.tabset}
```{r echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.width=14, fig.height=10}
for (sample.id in unique(s.obj$name)){
  cat(sprintf("### Sample %s \n", sample.id))
  p <- VlnPlot(object = subset(s.obj, name == sample.id), features = "AmbientRNA")
  print(p)
  cat("\n \n")
}
```

# Differential gene expression analysis: high ambient RNA contaminated cells vs others (all samples)

- positive logFC indicates that the gene is higher in the "ambient RNA contaminated cells >= 50%" cells. - negative logFC indicates that the gene is lower in the "ambient RNA contaminated cells >= 50%" cells.

```{r echo=FALSE, fig.height=10, fig.width=14, message=FALSE, warning=FALSE, results='asis'}
high.ambientRNA.cells <- subset(s.obj@meta.data, s.obj@meta.data$AmbientRNA >= 0.5) %>% row.names()

meta.data <- s.obj@meta.data %>% rownames_to_column("barcode") %>%
  rowwise() %>%
  mutate(high.ambientRNA = ifelse(barcode %in% high.ambientRNA.cells, "yes", "no")) %>%
  column_to_rownames("barcode")
meta.data <- meta.data[row.names(s.obj@meta.data), ]
s.obj <- AddMetaData(object = s.obj, metadata = meta.data$high.ambientRNA, col.name = "high.ambientRNA")

if (file.exists(file.path(path.to.04.output, "preprocessed_subcluster_obj", sprintf("high.ambientRNA.markers_%s.rds", sub.cluster.id))) == FALSE){
  high.ambientRNA.markers <- FindMarkers(object = s.obj, ident.1 = "yes", ident.2 = "no", group.by = "high.ambientRNA", assay = "RNA", test.use = "wilcox")
  high.ambientRNA.markers.raw <- high.ambientRNA.markers
  high.ambientRNA.markers <- subset(high.ambientRNA.markers, high.ambientRNA.markers$p_val_adj <= 0.05)
  saveRDS(high.ambientRNA.markers.raw, file.path(path.to.04.output, "preprocessed_subcluster_obj", sprintf("high.ambientRNA.markers_%s.raw.rds", sub.cluster.id)))
  saveRDS(high.ambientRNA.markers, file.path(path.to.04.output, "preprocessed_subcluster_obj", sprintf("high.ambientRNA.markers_%s.rds", sub.cluster.id)))
} else {
  high.ambientRNA.markers <- readRDS(file.path(path.to.04.output, "preprocessed_subcluster_obj", sprintf("high.ambientRNA.markers_%s.rds", sub.cluster.id)))
  high.ambientRNA.markers.raw <- readRDS(file.path(path.to.04.output, "preprocessed_subcluster_obj", sprintf("high.ambientRNA.markers_%s.raw.rds", sub.cluster.id)))
}

high.ambientRNA.markers <- high.ambientRNA.markers %>% rownames_to_column("Gene") %>% rowwise() %>%
  mutate(abs.avg_log2FC = abs(avg_log2FC)) %>% arrange(desc(abs.avg_log2FC))
```

## Full table: All differentially expressed genes
```{r results='asis', echo=FALSE, message=FALSE, warning=FALSE, fig.width=14, fig.height=10}
high.ambientRNA.markers %>% create_dt()
```

```{r results='asis', echo=FALSE, message=FALSE, warning=FALSE, fig.width=14, fig.height=10}
input.df <- high.ambientRNA.markers.raw %>% rownames_to_column("Gene")
all.pvals <- unique(input.df$p_val_adj)
all.pvals <- all.pvals[all.pvals != 0]
min.pval.adj <- min(all.pvals)
input.df <- input.df %>% rowwise() %>%
  mutate(p_val_adj = ifelse(p_val_adj == 0, min.pval.adj, p_val_adj))

top10up.genes <- input.df %>% subset(p_val_adj <= 0.05 & avg_log2FC > 0) %>% arrange(desc(avg_log2FC)) %>% head(10)
top10down.genes <- input.df %>% subset(p_val_adj <= 0.05 & avg_log2FC < 0) %>% arrange(desc(avg_log2FC)) %>% tail(10)
input.df <- input.df %>%
rowwise() %>%
mutate(sig = ifelse(p_val_adj <= 0.05, "Sig.", "not Sig.")) %>%
  mutate(show.gene.name = ifelse(Gene %in% c(top10up.genes$Gene, top10down.genes$Gene), Gene, NA))
  
cutoff.adjp <- 0.05
volcano.plot <- ggplot(data=input.df, aes(x=avg_log2FC, y=-log10(p_val_adj ), col = sig, label=Gene)) +
  geom_point() +
  geom_label_repel(label = input.df$show.gene.name, size = 8, max.overlaps = 100) +
  scale_color_manual(values=c("#c0d2f0", "#f28095")) +
  theme_minimal() +
  geom_vline(xintercept=c(-1, 1), col="#9a9fa6", linetype='dotted') +
  geom_hline(yintercept=-log10(cutoff.adjp), col="#9a9fa6", linetype='dotted') +
  theme_bw() +
  theme(plot.title = element_text(hjust=0.5, face="bold", size = 12), axis.text=element_text(size=12))
volcano.plot
```

# Differential gene expression analysis: high ambient RNA contaminated cells vs others (d4_LPS and d4_SPF samples only)
```{r echo=FALSE, fig.height=10, fig.width=14, message=FALSE, warning=FALSE, results='asis'}
s.obj <- subset(s.obj, name %in% c("d4_LPS", "d4_SPF"))

high.ambientRNA.cells <- subset(s.obj@meta.data, s.obj@meta.data$AmbientRNA >= 0.5) %>% row.names()

meta.data <- s.obj@meta.data %>% rownames_to_column("barcode") %>%
  rowwise() %>%
  mutate(high.ambientRNA = ifelse(barcode %in% high.ambientRNA.cells, "yes", "no")) %>%
  column_to_rownames("barcode")
meta.data <- meta.data[row.names(s.obj@meta.data), ]
s.obj <- AddMetaData(object = s.obj, metadata = meta.data$high.ambientRNA, col.name = "high.ambientRNA")

```

- positive logFC indicates that the gene is higher in the "ambient RNA contaminated cells >= 50%" cells. 

- negative logFC indicates that the gene is lower in the "ambient RNA contaminated cells >= 50%" cells.

## UMAP

```{r echo=FALSE, fig.height=10, fig.width=14, message=FALSE, warning=FALSE, results='asis'}
DimPlot(object = s.obj, reduction = "INTE_UMAP", label = TRUE, label.box = TRUE, repel = TRUE)
```

```{r echo=FALSE, fig.height=10, fig.width=14, message=FALSE, warning=FALSE, results='asis'}
if (file.exists(file.path(path.to.04.output, "preprocessed_subcluster_obj", sprintf("high.ambientRNA.markers_%s.d4_LPS_and_d4_SPF.rds", sub.cluster.id))) == FALSE){
  high.ambientRNA.markers <- FindMarkers(object = s.obj, ident.1 = "yes", ident.2 = "no", group.by = "high.ambientRNA", assay = "RNA", test.use = "wilcox")
  high.ambientRNA.markers.raw <- high.ambientRNA.markers
  high.ambientRNA.markers <- subset(high.ambientRNA.markers, high.ambientRNA.markers$p_val_adj <= 0.05)
  saveRDS(high.ambientRNA.markers.raw, file.path(path.to.04.output, "preprocessed_subcluster_obj", sprintf("high.ambientRNA.markers_%s.d4_LPS_and_d4_SPF.raw.rds", sub.cluster.id)))
  saveRDS(high.ambientRNA.markers, file.path(path.to.04.output, "preprocessed_subcluster_obj", sprintf("high.ambientRNA.markers_%s.d4_LPS_and_d4_SPF.rds", sub.cluster.id)))
} else {
  high.ambientRNA.markers <- readRDS(file.path(path.to.04.output, "preprocessed_subcluster_obj", sprintf("high.ambientRNA.markers_%s.d4_LPS_and_d4_SPF.rds", sub.cluster.id)))
  high.ambientRNA.markers.raw <- readRDS(file.path(path.to.04.output, "preprocessed_subcluster_obj", sprintf("high.ambientRNA.markers_%s.d4_LPS_and_d4_SPF.raw.rds", sub.cluster.id)))
}

high.ambientRNA.markers <- high.ambientRNA.markers %>% rownames_to_column("Gene") %>% rowwise() %>%
  mutate(abs.avg_log2FC = abs(avg_log2FC)) %>% arrange(desc(abs.avg_log2FC))
```

## Full table: All differentially expressed genes
```{r results='asis', echo=FALSE, message=FALSE, warning=FALSE, fig.width=14, fig.height=10}
high.ambientRNA.markers %>% create_dt()
```

## Volcano plot
```{r results='asis', echo=FALSE, message=FALSE, warning=FALSE, fig.width=14, fig.height=10}
input.df <- high.ambientRNA.markers.raw %>% rownames_to_column("Gene")
all.pvals <- unique(input.df$p_val_adj)
all.pvals <- all.pvals[all.pvals != 0]
min.pval.adj <- min(all.pvals)
input.df <- input.df %>% rowwise() %>%
  mutate(p_val_adj = ifelse(p_val_adj == 0, min.pval.adj, p_val_adj))

top10up.genes <- input.df %>% subset(p_val_adj <= 0.05 & avg_log2FC > 0) %>% arrange(desc(avg_log2FC)) %>% head(10)
top10down.genes <- input.df %>% subset(p_val_adj <= 0.05 & avg_log2FC < 0) %>% arrange(desc(avg_log2FC)) %>% tail(10)
input.df <- input.df %>%
rowwise() %>%
mutate(sig = ifelse(p_val_adj <= 0.05, "Sig.", "not Sig.")) %>%
  mutate(show.gene.name = ifelse(Gene %in% c(top10up.genes$Gene, top10down.genes$Gene), Gene, NA))
  
cutoff.adjp <- 0.05
volcano.plot <- ggplot(data=input.df, aes(x=avg_log2FC, y=-log10(p_val_adj ), col = sig, label=Gene)) +
  geom_point() +
  geom_label_repel(label = input.df$show.gene.name, size = 8, max.overlaps = 100) +
  scale_color_manual(values=c("#c0d2f0", "#f28095")) +
  theme_minimal() +
  geom_vline(xintercept=c(-1, 1), col="#9a9fa6", linetype='dotted') +
  geom_hline(yintercept=-log10(cutoff.adjp), col="#9a9fa6", linetype='dotted') +
  theme_bw() +
  theme(plot.title = element_text(hjust=0.5, face="bold", size = 12), axis.text=element_text(size=12))
volcano.plot
```

